File: problems/0647.0_Palindromic_Substrings.py
'''
中心扩散法
T： O（N * （N + N）） = O（N^2)
S: O(1)

Runtime: 165 ms, faster than 61.40% of Python3 online submissions for Palindromic Substrings.
Memory Usage: 13.9 MB, less than 75.82% of Python3 online submissions for Palindromic Substrings.
'''
class Solution:
    def countSubstrings(self, s: str) -> int:
        ans = 0
        n = len(s)
        
        for i in range(n):
            # odd
            step = 1
            # i - step ...i ... i + step
            while i - step >= 0 and i + step < n and s[i - step] == s[i + step]:
                step += 1
            ans += step
            
            # even
            step = 1
            # i - step + 1 ... i + step
            while i - step + 1 >= 0 and i + step < n and s[i - step + 1] == s[i + step]:
                step += 1            
            ans += step - 1
            
        return ans 

File: java/0647.0_Palindromic_Substrings.java
/*
中心扩散法
T： O（N^2)
S: O(1)

Runtime: 6 ms, faster than 60.83% of Java online submissions for Palindromic Substrings.
Memory Usage: 42 MB, less than 65.16% of Java online submissions for Palindromic Substrings.
*/
class Solution {
    int num = 0;
    public int countSubstrings(String s) {
        for (int i = 0; i < s.length(); ++i) {
            count(s, i, i);
            count(s, i, i + 1);
        }
        return num;
    }
    
    public void count(String s, int start, int end) {
        while (start >= 0 && end < s.length() && s.charAt(start) == s.charAt(end)) {
            num++;
            start--;
            end++;
        }
    }
}
